src/main/java/com/fcgl/madrid/points/service/UserActionService.java

```diff
@@ -1,22 +1,96 @@
 package com.fcgl.madrid.points.service;
 
-import com.fcgl.madrid.points.model.UserAction;
-import com.fcgl.madrid.points.repository.UserActionRepository;
+import com.fcgl.madrid.points.model.*;
+import com.fcgl.madrid.points.repository.*;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
 import org.springframework.stereotype.Service;
 
 import java.util.List;
 
 @Service
 public class UserActionService {
     private UserActionRepository userActionRepository;
+    private ActionRepository actionRepository;
+    private TrophyRepository trophyRepository;
+    private UserTrophyRepository userTrophyRepository;
+    private UserPointsRepository userPointsRepository;
 
     @Autowired
-    public UserActionService(UserActionRepository userActionRepository) {
+    public UserActionService(UserActionRepository userActionRepository,
+                             ActionRepository actionRepository,
+                             TrophyRepository trophyRepository,
+                             UserTrophyRepository userTrophyRepository,
+                             UserPointsRepository userPointsRepository) {
         this.userActionRepository = userActionRepository;
+        this.actionRepository = actionRepository;
+        this.trophyRepository = trophyRepository;
+        this.userTrophyRepository = userTrophyRepository;
+        this.userPointsRepository = userPointsRepository;
     }
 
     public List<UserAction> findAll() {
         return userActionRepository.findAll();
     }
+
+    public ResponseEntity<Trophy> addUserAction(Long userID, Long actionID) {
+
+        // Check that the action exists
+        Action action = actionRepository.findById(actionID).orElse(null);
+        if (action == null) {
+            return ResponseEntity.badRequest().body(null);
+        }
+
+        UserAction userAction = userActionRepository.findByUserIdAndAction(userID, action).orElse(null);
+        if (userAction == null) {
+            // We haven't seen this action for this user,
+            // Let's create a new entry with count equal to 1
+            userAction = new UserAction(userID, action, 1);
+        } else {
+            // We have seen this action for this user before
+            // just increment the action counter
+            userAction.setCount(userAction.getCount() + 1);
+        }
+        userActionRepository.save(userAction);
+
+        UserPoint userPoint = userPointsRepository.getByUserId(userID).orElse(null);
```

Comment:
Line 56 to Line 62 can be turned into a private function

Reply:
done!